### PR TITLE
Load the entities list using the view context

### PR DESCRIPTION
**Pull request by youknowriad**
See #37489 

Some users have access to "view" entities but not to "edit" theme. These users should be able to use the "core-data" package to fetch these entities. This is not possible sometimes in trunk because the way we "discover" entities is by making an "edit" context request to the `/types` endpoint. 

I think it makes more sense to use the "view" context to discover entities instead. That said, right now we rely on the "singular" label that is not returned in the "view" context. We have a number of options:

 - Accept that we use the "plural" one in some parts of the UI (not ideal)
 - Add the `singular label` (we already have the "name", we could add a "singlularName") or the entirely "labels" property to the returned value of the "view" context. Not sure if the REST team discussed this.

In this current PR, I'm trying the first approach to check the results.

### CODE REVIEW

**Comment by youknowriad** [comment]
If I'm not wrong, the label change (plural instead of singular) mostly impacts the multi-entity save panel where it's going to show things like "Posts" instead of "Post" as title. I think it's probably a fine tradeoff for an important fix. 

If this proves to be something folks are not ok with, we can consider finding a way to access singular labels for users without "edit" context rights.

WDYT?

**Comment by mcsf** [comment]
> If I'm not wrong, the label change (plural instead of singular) mostly impacts the multi-entity save panel where it's going to show things like "Posts" instead of "Post" as title. I think it's probably a fine tradeoff for an important fix.

Also: the name of the tab in the block sidebar (SettingsHeader) and the breadcrumbs.

The proposal makes sense to me, `view` seems like a more appropriate context. If the only caveat is the type labels, we should look at those. I found it surprising that they aren't part of the `view` context — I wonder if there is a real implication there, or if it was just a matter of [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

**Comment by youknowriad** [comment]
@adamziel @getdave would love a check on this one when you have time. Thank you.

**Comment by getdave** [comment]
Yeh sorry got distracted last week. Let me take a look!

File: packages/core-data/src/entities.js
Comment (getdave) [inline packages/core-data/src/entities.js:219]: This change is going to be difficult to test and the results could be unexpected (although likely relatively minor).

Ideally we would modify the Types API endpoint to provide access to the singular label as that will avoid this change cause knock on effects to UI labels (some of which may be handled in 3rd party plugins).

**Review by getdave — COMMENTED** [review]
In general I couldn't see any major UI bugs resulting from this change. However, I'm sure when I surfaced the issue (and when we chatted) there were other knock on effects for lower permission users such as allowing you to attempt edit requests to entities you cannot edit. Going to dig a little deeper.

The UI continued also continued to function for lower permission users.

## Context

I flagged this issue to Riad because I noticed that (on `trunk`) `getEntityRecord` will not allow for _reading_ CPTs even if the user has permission to view. 

This is because [the Types API endpoint only returns those posts with `edit` permission if you provide the `context=edit`](https://github.com/WordPress/wordpress-develop/blob/5cb75a136d1134cea62b2ccd595c0f7405060466/src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php#L83-L101). 

As a result this line of `getEntityRecord` will short circuit any request by lower permission users to read any post that is not returned by the Types endpoint (_even_ if you pass the `context=view` in your call to `getEntityRecord`):

https://github.com/WordPress/gutenberg/blob/c9ba45839c0e2521d507d95757ab1bfdb0a307ec/packages/core-data/src/resolvers.js#L55-L59

Note that the following line makes a request to the Types endpoint with the context set to `edit`.

https://github.com/WordPress/gutenberg/blob/c9ba45839c0e2521d507d95757ab1bfdb0a307ec/packages/core-data/src/resolvers.js#L55

This PR changes that so that we make the same request but using the `view` context. Therefore we get all _readable_ Types back which allows us to _read_ Post Types using `getEntityRecord`.


## Singular vs Plural Lables

@mcsf mentioned the labels in the multi-entity panel are now plural rather than singular. To me that doesn't seem so bad.

However, it could be good to amend the `<EntityTypeList>` component to use the singular / plural version of the Post label depending on the length of the results. That would however, require amending the Types API as suggested in my review comments. The `wp_template` already does this manually

https://github.com/WordPress/gutenberg/blob/5d9a7a72f7a3ef1c945b4d9ad8ac61d95b9fcc8d/packages/editor/src/components/entities-saved-states/entity-type-list.js#L52-L55

**Comment by spacedmonkey** [comment]
In my testing, I was getting a auth error see screenshot. 
Screenshot 2022-01-10 at 11 12 36

Steps to replicate. 
1. Login as a contributor. 
2. Create a post
3. Add a navigation block. 

The reason you are getting the auth error here, is the context of edit is being passed to endpoint where the user do not have access to the edit the entities. I noted all of this [here](https://github.com/WordPress/gutenberg/issues/37489#issuecomment-1001773451). 

This line needs to change 

https://github.com/WordPress/gutenberg/blob/812fa1b1f0d0015d9ba5a16ba17610250db08f46/packages/core-data/src/entities.js#L217

Based on if the user do / does not have access edit the entity or not. 

Solutions. 

1. Do one request `/wp/v2/types?context=edit` and another to `/wp/v2/types?context=view`. If the post type / taxonomy appears in edit response, set the default context to edit or fallback to view. 
2. Add a new field of default_context to the REST API for post type / taxonomy  and read these values in here. 
3. Do a options request on every post type / taxonomy to see if user has edit capaibility 

I recommend option 2. 

Willing to work on the PHP if required.

**Comment by youknowriad** [comment]
A new API to solve that particular navigation issue is not necessary, we can already pass "context: view" in specific `getEntityRecords` calls. The main issue we had before is that it was not possible because the "navigation" entity was not there at all (which the current PR solves)

**Comment by Mamaduka** [comment]
> I found it surprising that they aren't part of the view context — I wonder if there is a real implication there, or if it was just a matter of YAGNI.

I'm also curious about this.

@spacedmonkey, do you happen to know why labels are limited to `edit` context?

**Comment by spacedmonkey** [comment]
> A new API to solve that particular navigation issue is not necessary,

What new api are you talking about here? I don't see reference to a new api anywhere in this thread.

> @spacedmonkey, do you happen to know why labels are limited to edit context?

This data isn't public as it is only shown in the admin. But don't see the harm in showing in the view context.

**Comment by spacedmonkey** [comment]
@youknowriad Hey, I have created a breakout PR, with improved handling of context, if user does not have permission to access the post type or taxonomy. https://github.com/WordPress/gutenberg/pull/37838

**Comment by youknowriad** [comment]
> What new api are you talking about here? I don't see reference to a new api anywhere in this thread.

I'm talking about adding a "defaultContext" information. We don't really need it. We can just pass the context to the `getEntityRecords` call that is causing the auth error. @getdave should be able to tell us which one is that.

**Comment by getdave** [comment]
To summarize:

* Jonny says we're making API calls using the incorrect context based on what the user has permission to do. 
* This is due to the fact that we end up with `getEntityRecord` calls [defaulting to using the `edit` context](https://github.com/WordPress/gutenberg/blob/d7c7af335f6108df6db673d5876de976e3fa2c73/packages/core-data/src/entities.js#L145). 
* Moreover many contributors are probably not clear on what the context field does (I feel a Make Post coming on...). so they never alter it
* Riad says we have an API available now to allow us to pass the correct context to `getEntityRecords`. This is correct but is only part of the problem as more often that not we don't know which context the user has permission on which post type.
* [Jonny explains](https://github.com/WordPress/gutenberg/issues/37489#issuecomment-1001773451) that we need to have a default context per post type which should be determined _dynamically_ based on the response form the Types endpoint. This would ensure we default all API calls to the correct context based on the user's permissions.

I think Jonny is on to something here. Riad is correct that we can manually pass a `view` context but we should automatically default to the _correct context_ based on user permissions. This data should come from the Types endpoint rather than being hard coded. 

If a developer wants to override that on a case by case basis then that's still fine, but in general the data API should pick the correct defaults. 

> We can just pass the context to the `getEntityRecords` call that is causing the auth error. @getdave should be able to tell us which one is that.

I believe it would be this one

https://github.com/WordPress/gutenberg/blob/d7c7af335f6108df6db673d5876de976e3fa2c73/packages/block-library/src/navigation/use-navigation-entities.js#L46

**Comment by youknowriad** [comment]
"defaultContext" is very weird for me. What is the default context for any entity if not "view"? why does it need to be a property. 

I disagree with the fact that we don't know the context we want on the component level, we know that we want to show a list of pages and not edit them, so that should be "view" context, we know that we want to edit a page, that should be "edit" context...

> getMenus( ...menusParameters ),

so changing this to `context: 'view'` to that menuParameters should solve this issue. (in addition to the current PR)

**Comment by getdave** [comment]
> I disagree with the fact that we don't know the context we want on the component level

What I think you're saying is that as a developer we should know what type of request is being made - are we looking to read or write the data.

Currently we default to `edit` but if we defaulted all entities to `view` then that would ensure that developers are forced to set the context correctly based on what they are trying to achieve.

As a result, we wouldn't need to include the default context because we can simply default to the lowest permission (i.e. `view`) and that will be ok.

> so changing this to `context: 'view'` to that menuParameters should solve this issue. (in addition to the current PR)

Happy to spin that one up.

**Comment by youknowriad** [comment]
> Currently we default to edit but if we defaulted all entities to view then that would ensure that developers are forced to set the context correctly based on what they are trying to achieve.

yes, the problem here is that the data package has been built originally for the post editor (editing posts) so we defaulted to edit and now we're stuck with this forever (backward compatibility)

**Comment by getdave** [comment]
> and now we're stuck with this forever (backward compatibility)

...oh. So you're saying that in fact we cannot default to `view`.

Could we introduce some new APIs such as `getViewableEntityRecords` and `getEditableEntityRecords` which set the `context` parameter accordingly and then proxy to the main `getEntityRecords`?

**Comment by spacedmonkey** [comment]
Sadly we can't just default to using `view` for all calls.  For a couple of reason. 

1. Fields like title.raw and content.raw will not be returned response.
2. This is breaking change. 
3. There maybe other fields that developers are expecting in context=edit based responses. 

Not having title.raw or content.raw would break things down stream, as this at the fields that are needed to used to editor. Without these fields, you can not edit the raw values. Title and content returned rendered version, which is useless in editing context.

> I'm talking about adding a "defaultContext" information.

To be clear, this is not a new api, this is just a new field in two existing apis. See #37838. I too don't love adding a new field, but it is being done for backwards compatibility reasons. This way, all the entities that currect are listed with context=edit, remain the same and entities that user can only view, are added to this list with the context of view. This fixes the issue without breaking / changing very much. Please take the time to [test my PR ](https://github.com/WordPress/gutenberg/pull/37838)and understand why I have requested this change.

**Comment by youknowriad** [comment]
> Could we introduce some new APIs such as getViewableEntityRecords and getEditableEntityRecords which set the context parameter accordingly and then proxy to the main getEntityRecords?

I think the current workaround (pass "context": "view") when we only need "view" fields and we want access for contributors is good enough IMO

**Comment by spacedmonkey** [comment]
What about ?

> Fields like title.raw and content.raw will not be returned response.

**Comment by youknowriad** [comment]
> Fields like title.raw and content.raw will not be returned response.

The root API doesn't change `getEntityRecords`, if you call it without arguments, it will still default to `edit` to return all fields but for the navigation block where we only want to "read" menus we pass the extra context argument to switch to "view" context explicitly (which was supposed to work already btw if not the entity discovery relying on edit context and solved here)

**Comment by spacedmonkey** [comment]
@youknowriad If that is the case, then why am I seeing this error 

Screenshot 2022-01-10 at 11 12 36

**Comment by youknowriad** [comment]
@spacedmonkey I didn't touch the navigation blocks `getEntityRecords` call yet here, I'm just fixing the framework issue that was preventing us from making that change. I don't know that block enough to make the change myself. I'd prefer to leave this for folks already familiar with it like @getdave or @adamziel

**Comment by getdave** [comment]
~I'll spin up a PR~ [I've spun up a PR](https://github.com/WordPress/gutenberg/pull/37884) so we can talk about a specific example of this change in the context of the Nav block and fetching Menus.

It's difficult (at least for me) to discuss this in abstract so perhaps that will help. 

I will get to that shortly spinning off this PR as a base.

**Review by spacedmonkey — APPROVED** [review]

**Review by getdave — APPROVED** [review]
My instinct is that this is ok to merge _but_ we should follow up on the Types endpoint to provide full access to labels.

**Comment by noisysocks** [comment]
Hi! 👋 Friendly nudge that today (Monday) is the last day for this PR to be merged if it is to ship in WP 5.9. I'll begin backporting merged PRs on Tuesday morning Australian Eastern time.

**Comment by youknowriad** [comment]
I'm not entirely sure this needs backporting, I'll leave it to you all (Navigation block experts) but I feel like if we can wait it's better (due to the labels change).

**Comment by getdave** [comment]
> I'll leave it to you all (Navigation block experts)

Let me check this out now - I'll report back with a view shortly.

**Comment by getdave** [comment]
I don't think we should backport this PR. For two reasons.

Firstly, the problem it's intended to unblock is currently a non-starter because we now no longer allow lower permission users to create menus from classic Menus via the Navigation block. This is because that requires the ability to create `wp_navigation` posts which is something lower permissions users (intentionally) don't have the capability to do. There is https://github.com/WordPress/gutenberg/pull/37884 but that now just avoids a 403 error in the console rather than anything in the UI change. 

Secondly, this is a _potentially_ major change and we're in RC. We should probably avoid unless it's _critical_ - as Im sure you all agree.

So let's leave it out of WP 5.9.

**Pull request merged**
